File: 백준/Gold/1911. 흙길 보수하기/흙길 보수하기.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.Comparator;
import java.util.StringTokenizer;

public class Main {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        StringTokenizer st = new StringTokenizer(br.readLine());

        int N = Integer.parseInt(st.nextToken());
        int L = Integer.parseInt(st.nextToken());

        int[][] n_list = new int[N][2];

        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());
            n_list[i][0] = Integer.parseInt(st.nextToken());
            n_list[i][1] = Integer.parseInt(st.nextToken());
        }

        Arrays.sort(n_list, Comparator.comparingInt(a -> a[0]));

        int ans = 0;
        int range = 0;

        for (int i = 0; i < N; i++) {
            if (n_list[i][0] > range) {
                range = n_list[i][0];
            }

            if (n_list[i][1] > range) {
                while (n_list[i][1] > range) {
                    range += L;
                    ans++;
                }
            }
        }


        System.out.println(ans);
        br.close();

    }
}

File: 백준/Gold/1911. 흙길 보수하기/흙길 보수하기.py
N, L = map(int, input().split())

n_list = [list(map(int, input().split())) for _ in range(N)]

# 정렬
n_list.sort()

# 널빤지 수
ans = 0

for i in range(N):
    length = n_list[i][1] - n_list[i][0]

    # 마지막 웅덩이
    if i == N-1:
        ans += ((length-1) // L + 1)
        break
    # 길이가 나누어떨어지지 않을 경우
    if length % L:
        # 널빤지의 남은 부분
        temp_length = L - (length % L)
        # 다음 길이 비교해야 함
        _next = n_list[i][1] + temp_length

        if _next >= n_list[i+1][0]:
            n_list[i+1][0] = _next

        # 더해주기
        ans += (length // L + 1)
    else:
        ans += length // L

print(ans)


